povm_toolbox/post_processor/state_optimal.py

```diff
@@ -0,0 +1,49 @@
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""TODO."""
+
+from __future__ import annotations
+
+from qiskit.quantum_info import DensityMatrix, SparsePauliOp, Statevector
+
+from povm_toolbox.post_processor import POVMPostProcessor
+from povm_toolbox.quantum_info import MultiQubitPOVM
+from povm_toolbox.quantum_info.multi_qubit_dual import MultiQubitDUAL
+
+
+class StateOptimal(POVMPostProcessor):
+    """A POVM post-processor that leverages state knowledge to set the dual frame."""
+
+    def set_state_optimal_dual(
+        self,
+        state: SparsePauliOp | DensityMatrix | Statevector,
+    ) -> None:
+        """Set the dual frame based on the outcome distribution of a supplied state.
+
+        This methods constructs a joint dual frame where the alpha-parameters are
+        set as the outcome probabilities of the supplied state. It can be shown
+        that this is the dual frame minimizing the variance of the estimator (irrespective
+        of the observable to estimate) if outcomes are sampled from this state.
+
+        Args:
+            state: state from which to compute the outcome probabilities.
+
+        Raises:
+            NotImplementedError: if ``self.povm`` is not a :class:`povm_toolbox.quantum_info.product_povm.MultiQubitPOVM`
+                instance.
+        """
+        if not isinstance(self.povm, MultiQubitPOVM):
+            # TODO : implement for ProductPOVM
+            raise NotImplementedError(
+                "This method is only implemented for `povm_toolbox.quantum_info.product_povm.MultiQubitPOVM`."
+            )
+        alphas = tuple(self.povm.get_prob(state))  # type: ignore
+        self._dual = MultiQubitDUAL.build_dual_from_frame(self.povm, alphas=alphas)
```

Comment:
I think this is a prime candidate for a different implementation approach of the sort that we were discussing offline: rather than making subclasses of `POVMPostProcessor` which all just add a single method to set a custom `self._dual`, maybe we can expose a public setter for `self.dual` on the `POVMPostProcessor` and then provide a number of different functions which the user calls to set this dual.

Here, this would be a particularly simple example:
```python
def optimal_dual_from_state(state, povm):
    alphas = tuple(povm.get_prob(state))
    return MultiQubitDUAL.build_dual_from_frame(povm, alphas=alphas)

post_processor = POVMPostProcessor(result)
post_processor.dual = optimal_dual_from_state(state, povm)
```

Reply:
The same will apply to the other subclasses, but I will review them in terms of their content, not w.r.t. this design aspect.